[Ng Wei Chong Javier] iP

javierng2knus opened the pull request:
> [!NOTE]
> This is still currently a work in progress project.

```
      ____.  _________   ____
     |    | /  _  \   \ /   /
     |    |/  /_\  \   Y   /
 /\__|    /    |    \     /
 \________\____|__  /\___/
                  \/
```

# Jav the Joy Amplifying Virtuoso 🎷

<details><summary>Some of the features</summary>
<p>

1. Greetings & Farewells
2. Stores tasks for you
   - To-dos (Do X)
   - Deadlines (Complete X by Y)
   - Events (Attend X from Y to Z)

</p>
</details> 

   
## How to use it
- [ ] Download the files
- [ ] ???
- [ ] **_Profit_**

Design patterns that are used:

- Singleton pattern using `LazyHolder` from [wikipage](https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom)

```java
public class Something {
    private Something() {}

    private static class LazyHolder {
        static final Something INSTANCE = new Something();
    }

    public static Something getInstance() {
        return LazyHolder.INSTANCE;
    }
}
```

> [!TIP]
> Be sure to let me know if there are any room for improvement! I'm always eager to learn more!

> The only true wisdom is in knowing you know nothing - Socrates [(source)](https://www.goodreads.com/author/show/275648.Socrates)

> [!WARNING]
> This is done on windows, mac and linux users _might_ face potential issues.

> [!CAUTION]
> My code is not perfect so refer to it at your own risk!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

javierng2knus committed:
Modified gitignore for class files

Just to test commit and push

javierng2knus committed:
Added greeting functionality

Added greeting function for greeting user when starting the program.
Updated Duke to JAV

javierng2knus committed:
Added exit msg

Added exit function for say farewell to user when exiting the program.
Updated some text to be more reflective of the personality

javierng2knus committed:
Added echo and some fixes

Added Echo function to echo any message
Updated the printing format
Fixed some commenting

javierng2knus committed:
Added Add and List functionality

Added Storage class for storage needs
Added MessagePrinter class for message needs
Split functionality to classes so no more god class
Updated comments

javierng2knus committed:
Added marking and unmarking + fixes

Added functionality to mark and unmark
Added Task class
Modified Storage and MessagePrinter to work with Task class
More update to comments
Modified README to reflect additions

javierng2knus committed:
Added 3 types of tasks; todo, event, deadline

Added todo class
Added deadline class
Added event class
Adjusted main class for easier checking
Added enum for the different types of tasks

javierng2knus committed:
Added TextUiTesting

Modified input and expected txt with runtest batch file to semi-automate test for regression
Remove random tempCodeRunnerFile from vscode
Remove debugging text in event.java

javierng2knus committed:
Added exception handling

Added exception handling

javierng2knus committed:
Added remove function

Added remove function
Updated readme
Adjusted test for remove function

javierng2knus committed:
Added Save functionality

Added FileManager class for file reading and writing
Added save functionality
Updated task subclasses for for reading and saving
Updated runtest.bat to delete savefile when testing

javierng2knus committed:
Added time feature

Updated README to include save
Added time feature to Deadline class
Updated testing

javierng2knus committed:
Revert "Added time feature"

This reverts commit 75bb637a3f18cc7a50d8490ba3c52bceac20512d.

javierng2knus committed:
Added time feature

Updated README to include save
Added time feature to Deadline class
Updated testing

did an oppsie and didn't branch :(

javierng2knus committed:
Adhering to coding standard + more OOP

Modified everything so they fit the coding standard, rip time
Made the code more OOP
-Already have Ui (MessagePrinter)
-Already have Storage (FileManager)
-Added Parser
-Already have TaskList (StorageManager)

javierng2knus committed:
Even more OOP

Extracted commands into their separate own class
Converted managers to singleton

javierng2knus committed:
Added packaging

Put all classes in some package

javierng2knus committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

javierng2knus committed:
Added gradle & JUnit testing

Added gradle task testing
Some fixes for packaging.
Added Junit test class (just for testing, will be repurposing for later)

javierng2knus committed:
Add JUnit tests for Deadlines and Events

Add JUnit tests for Deadlines and Events to test the behavior of the respective classes.

There are cases where construction of a deadline or an event could result in a error due to the invalid parameters.

The addition of JUnit tests will allow for such bugs to be caught with greater ease.

javierng2knus committed:
Update gradle settings for shadowJar

Update gradle settings for shadowJar gradle plugin to package an application into an executable JAR file.

The plugin required certain settings in the build.gradle and the file directory in the fileManager class to be configured to be successful in building the JAR file.

javierng2knus committed:
Ensure all code have Javadoc

Add JavaDoc to areas without it.

The goal is to add header comments to all non-private classes/methods, and non-trivial private methods.

javierng2knus committed:
Tweak the code to comply with coding standard

Modify all present lines of code to comply with the given coding standard.

The goal is to tweak the code to comply with the given coding standard outlined here https://se-education.org/guides/conventions/java/index.html

javierng2knus committed:
Add Find command

Add new command child class - FindTaskCommand.

The goal is to give users a way to find a task by searching for a keyword.

The implementation ensures the OOP structure through the parser-command system.

javierng2knus committed:
Merge branch 'branch-A-CodingStandard'

javierng2knus committed:
Merge branch 'branch-A-JavaDoc'

chiralcentre commented on .vscode/settings.json:
Should the .vscode folder be removed from the repository? From my understanding, it is common practice to add the .vscode folder to the .gitignore file.

chiralcentre commented on docs/README.md:
I like the level of detail in the README, which explains how to use the bot clearly

chiralcentre commented on src/main/java/jav/Jav.java:
Should the Boolean variable isExiting be a local variable inside the main class instead? I was thinking it does not make sense for all instances of the Jav class to share the same value of isExiting.

chiralcentre commented on src/main/java/jav/command/DeleteTaskCommand.java:
Would it be better to have different error messages to differentiate between the case where an integer was not provided and the case when an integer was provided but the integer is not in the valid range? Currently, for both cases, your code gives the same error message.

chiralcentre commented on src/main/java/jav/command/UpdateTaskMarkCommand.java:
Would it be better to have different error messages to differentiate between the case where an integer was not provided and the case when an integer was provided but the integer is not in the valid range? Currently, for both cases, your code gives the same error message.

chiralcentre commented on src/main/java/jav/manager/FileManager.java:
Would it be better to use StringBuilder in the load method? String concatenation is quite inefficient.

chiralcentre commented on src/main/java/jav/task/Event.java:
Should the startDate and endDate attributes be changed to LocalDate objects, in accordance with Level 8 requirements? At the same time, should the toString() methods use String.format() instead of concatenation for greater clarity?

chiralcentre commented on src/main/java/jav/task/Task.java:
Should Task be an abstract class at this point? From my understanding, there is no command to add a Task at the current stage, and thus there shouldn't be a need to instantiate an instance of the Task object. At the same time, the toString() methods for the subclasses of Task share a lot of similar code; would it be better to extract the common parts and place them in a helper method in Task instead?

chiralcentre reviewed (commented):
LGTM, but there are a few issues that I have highlighted, which I hope will be of use to you.

javierng2knus committed:
Adjust sections based on feedback

Adjust sections of the code that were previously not as clear, change the way toString() method is handled for all task related classes and replace usage of concatenation in FileManager to be StringBuilder instead.

The goal is to be clearer in comments and error messages, extract common toString() code so the subclasses have fewer sections of repeated code and be more efficient in FileManager with the use of StringBuilder.

Thanks to chiralcentre for the useful feedback!

jocelyn-soh commented on src/main/java/jav/Jav.java:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        // Main program loop
+        while (!isExiting) {
```
Nit: I think that using -ing form for boolean variables maybe a little contradictory since it's supposed to be only true or false but -ing refers to somewhere in between true and false. Perhaps you can consider variable names like isComplete, isFinished, etc.

jocelyn-soh commented on src/main/java/jav/task/Event.java:
```diff
@@ -0,0 +1,94 @@
+    protected LocalDate startDate;
+
+    /**Date formatted end date. */
+    protected LocalDate endDate;
```
I like the naming of these 4 variables! I think it's very clear and concise.

jocelyn-soh commented on src/main/java/jav/command/UpdateTaskMarkCommand.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public class UpdateTaskMarkCommand extends TaskCommand { 
+    /** Whether the task is going to be marked or unmarked. */
+    private boolean isMarking;
```
Please refer to a similar comment about isExiting in Jav.java

jocelyn-soh reviewed (commented):
Code generally adheres to code quality guidelines. Other than a few nits, it LGTM! Good job!

javierng2knus committed:
Integrate Checkstyle to project

Integrate Checkstyle to the project with Gradle and refactor code that contained violations.

The goal is to have a tool to automatically check the code for any coding standard violations.

Note that following config files and how it affects the checks:
config/checkstyle/checkstyle.xml: Rules to follow.
config/checkstyle/suppressions.xml: Rules to suppress under in which files.

javierng2knus committed:
Integrate basic GUI

Integrate basic GUI via JavaFX.

The goal is to use the JavaFX technology to implement the GUI and replace the previous text-based UI.

Guide found here: https://se-education.org/guides/tutorials/javaFx.html

javierng2knus committed:
Refactor text UI to GUI

Refactor all relevant UI code such as the responses to commands to properly reflect in the new GUI.

The goal is to make use of the previously implemented JavaFX technology and convert all of the existing text UI to the new GUI.

javierng2knus committed:
Improve code quality

Improve code quality with addition of guard clauses and better naming.

The goal is to maximize readability by avoiding deep nesting with arrowhead-like shapes and making the happy path more prominent with the addition of guard clauses.
Additionally there is some modification to the vague naming done regarding Booleans.

javierng2knus committed:
Improve code quality further

Address the issues highlighted to improve code quality.

The goal is to update and improve the code further based
on the feedback received by the soc-se-bot, addressing all
issues regarding Header Comments.

javierng2knus committed:
Include assert to document important assumptions

Include Java's assert feature to document important
assumptions that should hold at various points in the code.

The goal is to use assertions to check and document
important assumptions about certain arguments.

Additionally, in the process of searching for locations to
include assertions. A location where assertion could be
used was found in the DeleteTaskCommand and
UpdateTaskMarkCommand classes but instead of using
assertions, using exception handling is determined to be
more suitable.

javierng2knus committed:
Merge pull request #2 from javierng2knus/branch-A-Assertions

Include assert to document important assumptions

javierng2knus committed:
Implement undo command as extension

Implement two new classes to give the ability to undo the
most recent command.

The goal is to keep track of the commands done by the
user and allow undoing of successful commands. All
commands need to implement the undo method however
only certain commands are required to actually have some
functionality.

Also, some refactoring was done on the Task class and its
child classes to allow for easier undoing and the file system
is properly integrated with the current GUI (It was not
saving and loading after the GUI changes).

javierng2knus committed:
Finalize the features

Finalize features, bug fix and add screenshot of Ui as Ui.png.

The goal is to finalize all feature, resolve the bug involving task marking not being saved to the file and prepare the program for the final version.

javierng2knus committed:
Update README.md

Update README.md to reflect the finalized product.

The goal is to ensure the chatbot name is stated clearly at the top of the User Guide and provide the reader with enough guidance to be able to use all important features of the chatbot.

Inclusion of descriptions regarding the Undo command and removal of depreciated information such as greetings and farewell.

javierng2knus committed:
Update README.md

javierng2knus committed:
Update README.md

javierng2knus committed:
Update README.md
